xla/backends/profiler/gpu/cupti_tracer.h

```diff
@@ -48,6 +72,207 @@ struct CuptiTracerOptions {
+  absl::Status CreatePmSamplerObject();
+
+  // Requires pm sampler object
+  absl::Status CreateCounterDataImage();
+
+  // Clean up
+  absl::Status DestroyCounterAvailabilityImage();
+  absl::Status DestroyProfilerHostObj();
+  absl::Status DestroyConfigImage();
+  absl::Status DestroyPmSamplerObject();
+  absl::Status UnInitializeProfilerAPIs();
+  absl::Status DestroyCounterDataImage();
+
+  std::vector<char const*> default_metrics_{
+    "sm__cycles_active.sum",
+    "sm__inst_executed_pipe_fmalite.sum",
+    "pcie__read_bytes.sum",
+    "pcie__write_bytes.sum"
+  };
+
+  public:
+  // Device information
+  int device_id_;
+
+  // PM sampling public configuration
+  struct PmSamplingConfig* config_;
+
+  // Creates host and sampler objects, all images
+  absl::Status CreateConfig();
+
+  // Requires config image, pm sampler object
+  absl::Status SetConfig();
+
+  // Requires pm sampler object
+  absl::Status StartSampling();
+
+  // Requires pm sampler object
+  absl::Status StopSampling();
+
+  // Requires pm sampler object
+  absl::Status DisableSampling();
+
+  // Collect sampling data
+  // Requires pm sampler object, counter data image, fetches data from hw
+  // buffer into counter data image
+  absl::Status FillCounterDataImage(struct PmSamplingDecodeInfo& decodeInfo);
+
+  // Requires counter data image
+  absl::Status GetSampleCounts(struct PmSamplingDecodeInfo& decodeInfo);
+
+  // Requires host object, pm sampler object, counter data image, metric
+  // names, returns sample time, metric values
+  absl::Status GetSample(SamplerRange& sample, size_t index);
+
+  // Requires pm sampler object, counter data image, (re)initializes it
+  absl::Status InitializeCounterDataImage();
+
+  // Restores image from backup (faster than re-initializing)
+  absl::Status RestoreCounterDataImage();
+
+  // Constructor provides all configuration needed to set up sampling on a
+  // single device
+  PmSamplingDevice(int device_id, struct PmSamplingConfig* config);
+};
+
+// Container for PM sampling decode thread
+// Responsible for fetching PM sampling data from device and providing it to
+// handler or other external container
+class PmSamplingDecodeThread {
+  enum thd_state {
+    // Thread is starting, not yet ready to be enabled
+    state_uninitialized_,
+    // Thread is ready for enablement but decoding has not yet been triggered
+    state_initialized_,
+    // Thread is enabled, polling for metrics from all devices
+    state_enabled_,
+    // Thread is disabled, not polling for metrics, but could be re-enabled
+    state_disabled_,
+    // Thread is finishing and guaranteed to return, allowing join
+    state_exiting_
+  };
+
+  // Current state of the thread (only set by worker thread)
+  volatile thd_state current_thd_state_ = state_uninitialized_;
+    
+  // State thread should transition to (only set by main thread)
+  volatile thd_state nextThdState_ = state_initialized_;
+
+  // Thread handle
+  std::thread* thd_;
```

Comment:
Looking at initialisation (i.e. `thd_ = new std::thread(ThdFunc, this)`), it looks a memory leak. Use `std::unique_ptr<std::thread>` instead?